cli/cli.py

```diff
@@ -0,0 +1,282 @@
+
+    if args.debug:
+        enable_debug_logs()
+
+    # We need it for the commands
+    subparsers = parser.add_subparsers(help="sub-command help")
+
+    studio_cli = StudioCli(args.config)
+    studio_cli.build_commands(subparsers)
+
+    if not unprocessed_args:
+        parser.print_help()
+        sys.exit(1)
+
+    args = parser.parse_args()
+
+    command = unprocessed_args[0]
+
+    response = studio_cli.execute(command, args)
+
+    print(response)
+
+
+class StudioCli:
+    def __init__(self, config):
+        self.public_api_client = PublicAPIClient(config)
+        self.schema = self._get_schema()
+        self.commands = {}
+
+    def build_commands(self, subparsers):
+        for path, methods in self.schema["paths"].items():
+            for method, data in methods.items():
+                command = Command(path, method, data, self.public_api_client)
+                if self._is_valid_command(command):
+                    self.commands[command.name] = command
+                    self._extend_subparsers(command, subparsers)
+
+    def execute(self, command_name, args):
+        command = self.commands[command_name]
+        return command.execute(args)
+
+    # https://tw.instructuremedia.com/api/public/apidocs
+    def _get_schema(self):
+        response = self.public_api_client.request("get", "apidocs", version_prefix="")
+
+        if response.status_code != 200:
+            raise Exception(
+                f"Could not get details for {response.url}: {response.text}"
+            )
+
+        return response.json()
+
+    _type_to_type = {
+        "integer": int,
+        "string": str,
+        # not sure about these
+        "object": str,
+        "array": str,
+        "file": str,
+    }
+
+    def _is_valid_command(self, command):
+        missing_params = [
+            parameter
+            for parameter in command.path_params
+            if parameter not in command.parameter_names()
+        ]
+
+        if missing_params:
+            sys.stderr.write(
+                f"Skipping {command.name} for {command.path}, params not defined: {missing_params}\n"
+            )
+            return False
+
+        if any([param["in"] in ["body", "formData"] for param in command.parameters()]):
+            # commands in body parameters are not supported atm.
+            # same for file/formData parameters
+            return False
+
+        return True
+
+    def _extend_subparsers(self, command, subparsers):
+        command_parser = subparsers.add_parser(command.name, help=command.summary())
+
+        for param in command.parameters():
+            command_parser.add_argument(
+                f"--{param['name']}",
+                type=StudioCli._type_to_type[param.get("type", "string")],
+                required=param.get("required", False),
+                choices=param.get("enum", None),
+                help=param["description"],
+            )
+
+
+class Command:
+    _method_to_command = {
```

Comment:
Sorry to nitpick, but this is also private, so let's move it to the top of private stuff (right above `_is_singular`).

Reply:
👍